File: cpp/0905.0_Sort_Array_By_Parity.cpp
/*
执行用时：12 ms, 在所有 C++ 提交中击败了36.76% 的用户
内存消耗：15.6 MB, 在所有 C++ 提交中击败了96.45% 的用户
通过测试用例：285 / 285
*/
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
        partition(nums.begin(), nums.end(), [&](int x) {
            return (x & 1) == 0;
        });
        return nums;
    }
};


/*
执行用时：8 ms, 在所有 C++ 提交中击败了72.75% 的用户
内存消耗：15.8 MB, 在所有 C++ 提交中击败了60.21% 的用户
通过测试用例：285 / 285
*/
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
        partition(nums.begin(), nums.end(), [&](int x) {
            return (x % 2) == 0;
        });
        return nums;
    }
};


/*
执行用时：4 ms, 在所有 C++ 提交中击败了96.19% 的用户
内存消耗：15.8 MB, 在所有 C++ 提交中击败了67.04% 的用户
通过测试用例：285 / 285
*/
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
        stable_partition(nums.begin(), nums.end(), [&](int x) {
            return (x % 2) == 0;
        });
        return nums;
    }
};

File: problems/0905.0_Sort_Array_By_Parity.py
'''
Runtime: 102 ms, faster than 53.59% of Python3 online submissions for Sort Array By Parity.
Memory Usage: 14.7 MB, less than 17.91% of Python3 online submissions for Sort Array By Parity.
'''
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        return [x for x in nums if x % 2 == 0] + [x for x in nums if x % 2]


'''
sort

Runtime: 84 ms, faster than 81.12% of Python3 online submissions for Sort Array By Parity.
Memory Usage: 14.6 MB, less than 63.70% of Python3 online submissions for Sort Array By Parity.
'''
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        nums.sort(key=lambda x: x % 2 != 0)
        return nums


'''
sort

Runtime: 91 ms, faster than 67.73% of Python3 online submissions for Sort Array By Parity.
Memory Usage: 14.6 MB, less than 63.70% of Python3 online submissions for Sort Array By Parity.
'''
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        nums.sort(key=lambda x: 1 if x % 2 != 0 else 0)
        return nums

'''
sort 

Runtime: 139 ms, faster than 20.62% of Python3 online submissions for Sort Array By Parity.
Memory Usage: 14.7 MB, less than 63.70% of Python3 online submissions for Sort Array By Parity.
'''
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        nums.sort(key=lambda x: 12342 if x % 2 != 0 else 324)
        return nums

'''
sort

Runtime: 79 ms, faster than 90.70% of Python3 online submissions for Sort Array By Parity.
Memory Usage: 14.7 MB, less than 63.70% of Python3 online submissions for Sort Array By Parity.
'''
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        nums.sort(key=lambda x: 32 if x % 2 != 0 else -40)
        return nums


